src/backend.cpp

```diff
@@ -84,45 +93,37 @@ void Backend::resolve( const ResolverProblem& problem, ProblemSolutionList& todo
 	solProposals.append( QString::fromStdString( solPtr->description() ) );
     }
     
-    // Emit signal displayProblemSolution( QString, QStringList ) - First string = problem, second stringlist = solutions
+    // Emit signal displayProblemSolution( QString, QString ) - First string = problem, second string = solutions
     emit displayProblemAndSolutions( problemStatement, solProposals );
-    
-    // get the solutionID here
-    int solId;
-    while ( ( solId = m_oci->solutionID() ) == -1 );
-    
+}
+
+void Backend::applySolution( int solId )
+{
+    const ProblemSolutionList & solutions = m_currentProblem.solutions();
     ProblemSolutionList::const_iterator solIter = solutions.begin();
     std::advance( solIter, solId );
+    m_solutionsToTry.append( *solIter );
     
     qDebug() << "Returned solId from OCI - " << solId;
     qDebug() << "===== Selected solution =====";
     qDebug() << QString::fromStdString( ( *solIter )->description() );
     qDebug() << "=============================";
 	    
-    todo.push_back( *solIter );
+    if ( !m_resolverProblemList.isEmpty() )
+	resolve( *m_resolverProblemList.takeFirst() );
+    else {
+	qDebug() << "Apply selected solutions...";
+	m_zypp->resolver()->applySolutions( m_solutionsToTry.toStdList() );
+	qDebug() << "Dependencies solved";
+	resolveConflicts();
+    }
 }
 
 void Backend::resolve()
 {
-    ZYpp::Ptr zypp = ZypperUtils::zyppPtr();
-    
-    const ResolverProblemList & problems( zypp->resolver()->problems() );
-    ProblemSolutionList toTry;
-    
-    // list all the problems with solutions on the GUI
-    for ( const auto & probPtr : problems ) {
-	resolve( *probPtr, toTry );
-    }
-    
-    //apply solutions...
-    if ( !toTry.empty() ) {
-	qDebug() << "Apply selected solutions...";
-	zypp->resolver()->applySolutions( toTry );
-	qDebug() << "Dependencies solved";
-	return;
-    }
-    else 
-	throw "Solving dependencies failed: Giving up!";
+    m_resolverProblemList = QList<ResolverProblem_Ptr>::fromStdList( m_zypp->resolver()->problems() );
```

Comment:
m_solutionsToTry->clear(); is needed here in case resolve() is called two or more times. m_resolverProblemList is replaced with the new list but m_solutionsToTry isn't.